Outline rough interface in README.md

chrismytton opened the pull request:
This is some basic documentation for the `ScrapedPage` class that we prototyped in https://github.com/everypolitician-scrapers/guernsey-2016/pull/1 and https://github.com/everypolitician-scrapers/mongolia-khurai-wp-multiple-terms/pull/7.

chrismytton committed:
Outline rough interface in README.md

tmtmtmtm commented on README.md:
```diff
@@ -1,2 +1,30 @@
+  end
+end
+
+puts ExamplePage.new('http://example.com').to_h
```
I guess one key decision we need to make early is whether the URL is a bare parameter like this, or whether we go for a named parameter for it.

tmtmtmtm commented on README.md:
```diff
@@ -1,2 +1,30 @@
+
+class ExamplePage < ScrapedPage
+  field :title do
+    noko.at_css('h1').text
```
We probably want some explanation somewhere regarding what `noko` is here and where it comes from (not necessarily right now, but noting that this will be a stumbling point.) I guess it's a little harder to insert a TODO: into a README than into code…

chrismytton replied:
This makes me think that `noko` is possibly exposing too much of the implementation details. Perhaps we could expose the nokogiri methods in the page class directly?

I've opened https://github.com/everypolitician/scraped_page/issues/2 to track this so lets continue the discussion over there ➡️

tmtmtmtm commented on README.md:
````diff
@@ -1,2 +1,30 @@
+The default strategy for retrieving pages is to make an http request to retrieve them. If you'd also like to archive a copy of the page you're scraping into a git branch, you can pass a `:strategy` option to the constructor:
+
+```ruby
+ExamplePage.new('http://example.com', strategy: ScrapedPage::LiveRequestArchiveStrategy)
````
the implies we're going to pass bare classnames here? Just checking that that's a deliberate decision… (I think that'd probably also be better as ScrapedPage::Strategy::LiveRequestArchive — giving them all their own namespace)

chrismytton replied:
Hmm, now you mention it perhaps it would be better to fold this into a datastructure, so we'd have something more like this:

```ruby
ExamplePage.new('http://example.com', strategy: :live_request_archive)

# Somewhere else
strategies = {
  live_request_archive: ScrapedPage::Strategy::LiveRequestArchive,
}
```

And yes, giving them their own namespace seems like a very sensible idea.

chrismytton replied:
I suppose if scrapers provide their own strategies then it's best for them to work in the same way as the built in classes.

The reason I've done it this way is because when doing the prototype in the guernsey scraper I [needed to pass the `url` to the `strategy`](https://github.com/everypolitician-scrapers/guernsey-2016/blob/1f76f009d6dc487141b918f618aa3ed0dc663103/lib/scraped_page.rb#L14) and passing in a class seemed like the most straightforward way to do that. This does suggest that I've perhaps designed the interface to suit my needs as an implementor rather than a user though 🤔

chrismytton committed:
Remove trailing whitespace

chrismytton committed:
Move LiveRequestArchive into a Strategy namespace

chrismytton committed:
Pass in an instance of a strategy to ScrapedPage subclasses

Having an instance rather than a class means that the user can configure
the strategy before they pass it in, e.g. they might want to set the
strategy to use javascript (i.e. capybara).

chrismytton committed:
Expand the initial example in README

This illustrates that the data is available as methods as well as using
the `#to_h` method.

chrismytton committed:
Add details about archiving and custom strategies

This describes strategies in more depth and gives an example of creating
your own strategy.

chrismytton commented on README.md:
````diff
@@ -1,2 +1,70 @@
+
+```ruby
+class FilesystemStrategy
+  def get(url)
````
I don't think `get` is the best name for this method. At the moment we're (purposely) not providing any way for people to make a `POST` request, they'd need to make their own custom strategy. But having a strategy that makes a `POST` request in a method called `get` seems a bit wrong.

I toyed with using `request` as the method name, but then in lots of Ruby libraries that has the method signature `request(method = :get, url)`.

Perhaps something like `fetch`?

tmtmtmtm replied:
Perhaps make it a noun rather than a verb?

chrismytton commented on README.md:
```diff
@@ -1,2 +1,70 @@
+class FilesystemStrategy
+  def get(url)
+    body = File.read(Digest::SHA1.hexdigest(url) + '.html')
+    ScrapedPage::Response.new(body: body)
```
I'm wondering if this requires the strategy to have too much information about other classes? Would it be better if this just returned a data structure and then the consuming class could turn that into a concrete instance?

tmtmtmtm replied:
Well, the current documentation of "A strategy is an object that responds to a `get` method and returns a `ScrapedPage::Response`" raises quite a few questions along how I know what a ScrapedPage::Response is and how to generate one etc. So either we need to document _that_ more, or change what it returns. I don't really understand deeply enough what the Response class is/does atm, so don't have much a view on that yet…

chrismytton commented:
@tmtmtmtm Would be good to get your thoughts on the updated interface.

tmtmtmtm commented on README.md:
```diff
@@ -1,2 +1,70 @@
+
+## Custom strategies
+
+If for some reason you can't scrape a site using one of the build in strategies
```
typo "build" → "built"

chrismytton replied:
✅

chrismytton committed:
Fix type build => built

Every.Single.Time.

tmtmtmtm commented on README.md:
```diff
@@ -1,2 +1,70 @@
+
+## Usage
+
+Create a subclass of `ScrapedPage` for each page you wish to scrape.
```
This is a bit ambiguous. If I'm going to scrape 500 MP pages, I presumably don't need 500 subclasses…

chrismytton replied:
I've updated this to be clearer that it's for each _type_ of page and given an example as well. https://github.com/everypolitician/scraped_page/pull/1/commits/d62eb3c33da06a1ae7598be1d856680195a11926

chrismytton committed:
Rename get => response

Use a noun rather than a verb for the method that strategies have to
implement. This makes it more obvious what the strategy is expected to
return.

chrismytton committed:
Strategies should return a Hash not a ScrapedPage::Response

Rather than expecting people who are creating strategies to know/care
about internal classes we simple ask them to return a Hash with a known
set of keys.

chrismytton committed:
Make README Usage introduction less ambiguous

Give an example of the kinds of subclass people might create.

chrismytton commented on README.md:
````diff
@@ -1,2 +1,73 @@
+```ruby
+class FilesystemStrategy
+  def response(url)
+    { body: File.read(Digest::SHA1.hexdigest(url) + '.html') }
````
I've changed the strategy entry point method to `response` rather than `get`. As you say a noun works better and is clearer here.

> Well, the current documentation of "A strategy is an object that responds to a get method and returns a ScrapedPage::Response" raises quite a few questions along how I know what a ScrapedPage::Response is and how to generate one etc. So either we need to document that more, or change what it returns. I don't really understand deeply enough what the Response class is/does atm, so don't have much a view on that yet…

Yep, I think I was exposing too much of the guts of the library. I've changed this to just be a `Hash` for now.

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm commented:
I haven't compared this to any implementation anywhere, but this all seems to makes to sense to me! 👍

chrismytton commented:
I'm ignoring Travis for now and merging anyway. It's failing because there isn't any code to test :)
